libpod/runtime_pod_pause_linux.go

```diff
@@ -0,0 +1,234 @@
+	podInfraRootfs := filepath.Join(root, "graph/vfs/pause")
+	g.SetRootPath(filepath.Join(podInfraRootfs, "rootfs"))
+	g.SetRootReadonly(true)
+	g.SetProcessArgs([]string{pauseCmd})
+
+	containerName := p.Name() + "-infra"
+
+	// Create the new container
+	ctr, err := newContainer(g.Config, r.lockDir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting new container")
+	}
+	ctr.config.StopTimeout = CtrRemoveTimeout
+
+	// Set namespace based on current runtime namespace
+	// Do so before options run so they can override it
+	if r.config.Namespace != "" {
+		ctr.config.Namespace = r.config.Namespace
+	}
+
+	var options []CtrCreateOption
+	options = append(options, r.WithPod(p))
+	options = append(options, WithRootFSFromImage(imgID, imgName, false))
+	options = append(options, WithName(containerName))
+	for _, option := range options {
+		if err := option(ctr); err != nil {
+			return nil, errors.Wrapf(err, "error running container create option")
+		}
+	}
+
+	ctr.valid = true
+	ctr.state.State = ContainerStateConfigured
+	ctr.runtime = r
+
+	if r.config.CgroupManager != SystemdCgroupsManager && r.config.CgroupManager != CgroupfsCgroupsManager {
+		return nil, errors.Wrapf(ErrInvalidArg, "unsupported CGroup manager: %s - cannot validate cgroup parent", r.config.CgroupManager)
+	}
+
+	// Require use of Pod Cgroup
+	podCgroup, err := p.CgroupPath()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error retrieving pod %s cgroup", p.ID())
+	}
+	ctr.config.CgroupParent = podCgroup
+	// set it as a pause container
+	ctr.config.IsPause = true
+
+	//TODO generalized
+	ctr.config.ShmDir = "/dev/shm"
+
+	// Set up storage for the container
+	if err := ctr.setupStorage(ctx); err != nil {
+		return nil, errors.Wrapf(err, "error setting up storage for pause container")
+	}
+	defer func() {
+		if err != nil {
+			if err2 := ctr.teardownStorage(); err2 != nil {
+				logrus.Errorf("Error removing partially-created container root filesystem: %s", err2)
+			}
+		}
+	}()
+
+	ctr.config.LogPath = filepath.Join(ctr.config.StaticDir, "ctr.log")
+	// Lock the pod to ensure we can't add containers to pods
+	// being removed
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if err := r.state.AddContainerToPod(p, ctr); err != nil {
+		return nil, err
+	}
+	return ctr, nil
+}
+
+// CreateAndStartPauseContainer wrap buildah new, buildah add, and libpod run
+// to create a pause container for a pod. A pause container becomes the
+// basis for kernel namespace sharing between containers in the pod.
+func (r *Runtime) CreateAndStartPauseContainer(ctx context.Context, p *Pod) error {
+	if !r.valid {
+		return ErrRuntimeStopped
+	}
+
+	newImage, err := r.getPauseImage(ctx, p)
+	if err != nil {
+		return err
+	}
+
+	data, err := newImage.Inspect(ctx)
+	if err != nil {
+		return err
+	}
+	imageName := newImage.Names()[0]
+	imageID := data.ID
+
+	ctr, err := r.getPauseContainer(ctx, p, imageName, imageID)
+	if err != nil {
+		return err
+	}
+
+	return ctr.Start(ctx)
```

Comment:
No need to start it here - wait until someone asks to start the pod

Reply:
If someone does a podman run --pod <pod> and the pause container isn't started, it will fail on the start. I guess the question is, if the pause container isn't running, should we start the pause container if someone does a podman run --pod?

Reply:
I want to add a `--recursive` option to start (and probably run, for situations like this) soon, but we should do it separately - for now, I think it's fine to say that you should only use pod commands on a pod when possible